File: ios/SampleUI/Sources/SampleUI/Model/UseSmileIDSampleLicenses.swift
import Foundation

/// The generated third-party notices, as the screen renders them.
public struct UseSmileIDSampleLicenses: Equatable {
  public let components: [UseSmileIDSampleNotice]

  public init(components: [UseSmileIDSampleNotice] = []) {
    self.components = components
  }

  public var isEmpty: Bool {
    components.isEmpty
  }
}

/// One component a partner ships, and what it is licensed under.
///
/// The text is the component's own, holder line included, so it belongs to the notice rather than
/// to a table keyed by licence: two MIT components carry two different copyright holders.
public struct UseSmileIDSampleNotice: Equatable, Decodable {
  public let component: String
  public let version: String
  public let licenseId: String
  public let licenseName: String
  /// Nil when the licence is known but its text ships inside the component instead of with us.
  public let text: String?
  /// Set only alongside a nil text: the page that carries the licence we cannot ship.
  public let url: String?

  public init(
    component: String,
    version: String,
    licenseId: String,
    licenseName: String,
    text: String?,
    url: String? = nil
  ) {
    self.component = component
    self.version = version
    self.licenseId = licenseId
    self.licenseName = licenseName
    self.text = text
    self.url = url
  }

  /// The version where there is one; a vendored notice inside another component carries none.
  public var subtitle: String {
    version.isEmpty ? licenseName : "\(version) · \(licenseName)"
  }
}

extension UseSmileIDSampleLicenses {
  /// Reads the generated notices out of the library bundle; an absent or unreadable asset is the
  /// empty state rather than a crash, because it means the build did not ship them.
  public static func bundled() -> UseSmileIDSampleLicenses {
    // `.module` is internal, so it cannot be this public method's default argument.
    bundled(in: .module)
  }

  static func bundled(in bundle: Bundle) -> UseSmileIDSampleLicenses {
    guard let url = bundle.url(forResource: "licenses", withExtension: "json"),
          let data = try? Data(contentsOf: url)
    else {
      return UseSmileIDSampleLicenses()
    }
    return decode(data)
  }

  static func decode(_ data: Data) -> UseSmileIDSampleLicenses {
    guard let asset = try? JSONDecoder().decode(Asset.self, from: data) else {
      return UseSmileIDSampleLicenses()
    }
    return UseSmileIDSampleLicenses(components: asset.components)
  }

  private struct Asset: Decodable {
    let components: [UseSmileIDSampleNotice]
  }
}

File: flutter/sample_ui/lib/src/model/use_smileid_sample_licenses.dart
import 'package:flutter/foundation.dart';

/// One component a partner ships, and the licence text it carries.
@immutable
class UseSmileIDSampleNotice {
  /// [licenseName] is null when no known signature matched, which is a label the screen omits
  /// rather than a licence it guesses at.
  const UseSmileIDSampleNotice({
    required this.component,
    required this.text,
    this.licenseId,
    this.licenseName,
  });

  /// The package or engine dependency the notice belongs to.
  final String component;

  /// The component's own licence text, verbatim.
  final String text;

  /// The SPDX id where a signature matched.
  final String? licenseId;

  /// The licence's display name, shown beneath the component.
  final String? licenseName;
}

/// The third-party notices, as the screen renders them.
@immutable
class UseSmileIDSampleLicenses {
  /// [components] is already sorted and carries no first-party entry.
  const UseSmileIDSampleLicenses({
    this.components = const <UseSmileIDSampleNotice>[],
  });

  /// Every notice the app ships.
  final List<UseSmileIDSampleNotice> components;

  /// Whether the notices reached this build at all.
  bool get isEmpty => components.isEmpty;

  /// Reads the notices Flutter's own build step collected into the bundle.
  ///
  /// The toolchain regenerates them from the resolved graph on every build, so unlike a committed
  /// asset they cannot go stale, and they already cover the engine's C++ dependencies.
  static Future<UseSmileIDSampleLicenses> bundled({
    Stream<LicenseEntry>? licenses,
  }) async => from(await (licenses ?? LicenseRegistry.licenses).toList());

  /// Groups entries by component, because one entry can name several packages and one package can
  /// appear in several entries.
  static UseSmileIDSampleLicenses from(Iterable<LicenseEntry> entries) {
    final Map<String, StringBuffer> byComponent = <String, StringBuffer>{};
    for (final LicenseEntry entry in entries) {
      final String text = entry.paragraphs
          .map((LicenseParagraph paragraph) => paragraph.text.trim())
          .where((String it) => it.isNotEmpty)
          .join('\n\n');
      if (text.isEmpty) {
        continue;
      }
      for (final String package in entry.packages) {
        if (_isFirstParty(package)) {
          continue;
        }
        final StringBuffer buffer = byComponent.putIfAbsent(
          package,
          StringBuffer.new,
        );
        if (buffer.isNotEmpty) {
          buffer.write('\n\n');
        }
        buffer.write(text);
      }
    }
    final List<String> names = byComponent.keys.toList()..sort();
    return UseSmileIDSampleLicenses(
      components: <UseSmileIDSampleNotice>[
        for (final String name in names)
          _notice(name, byComponent[name]!.toString()),
      ],
    );
  }
}

/// Licensed from Smile ID rather than a third party; `sample_ui`'s own identity is the same case.
bool _isFirstParty(String package) =>
    package.startsWith('usesmileid') || package == 'sample_ui';

UseSmileIDSampleNotice _notice(String component, String text) {
  final String lowered = text.toLowerCase().split(RegExp(r'\s+')).join(' ');
  for (final _Signature signature in _signatures) {
    if (signature.needles.every(lowered.contains)) {
      return UseSmileIDSampleNotice(
        component: component,
        text: text,
        licenseId: signature.id,
        licenseName: signature.name,
      );
    }
  }
  return UseSmileIDSampleNotice(component: component, text: text);
}

/// A licence recognisable from its own wording.
@immutable
class _Signature {
  const _Signature(this.id, this.name, this.needles);

  final String id;
  final String name;
  final List<String> needles;
}

/// Ordered: the first match wins, so a licence naming another inside its text cannot claim it.
const List<_Signature> _signatures = <_Signature>[
  _Signature('MPL-2.0', 'Mozilla Public License 2.0', <String>[
    'mozilla public license version 2.0',
  ]),
  _Signature('Apache-2.0', 'Apache License, Version 2.0', <String>[
    'apache license',
    'version 2.0',
  ]),
  _Signature('APSL-2.0', 'Apple Public Source License 2.0', <String>[
    'apple public source license',
  ]),
  // The third clause is the only thing separating these two, so it is what the match keys on.
  _Signature('BSD-3-Clause', 'BSD 3-Clause License', <String>[
    'redistributions in binary form',
    'endorse or promote products derived',
  ]),
  _Signature('BSD-2-Clause', 'BSD 2-Clause License', <String>[
    'redistributions in binary form',
  ]),
  _Signature('MIT', 'MIT License', <String>[
    'permission is hereby granted, free of charge',
  ]),
];
